history/subscriber/kafkaSubscriber.py

```diff
@@ -0,0 +1,236 @@
+        """
+            Handles a given kafka received message
+            :param message The message that has been received
+        """
+        raise NotImplementedError("Abstract method called")
+
+class TenancyHandler(KafkaEventHandler):
+
+    def __init__(self):
+        pass
+
+    @staticmethod
+    def _spawn_watcher(service, subject, handler):
+        topic = get_topic(service, subject)
+        if topic is None:
+            # TODO how to handle?
+            return
+        watcher = KafkaListener(topic, handler)
+        watcher.start()
+        return watcher
+
+    def tenants_bootstrap(self):
+        LOGGER.debug('will bootstrap')
+        target = "{}/admin/tenants".format(settings.AUTH)
+        response = requests.get(target, timeout=3)
+        if 200 <= response.status_code < 300:
+            payload = response.json()
+            LOGGER.debug('Got list of tenants', payload)
+            for tenant in payload['tenants']:
+                LOGGER.debug('initializing tenant %s', tenant)
+                TenancyHandler._spawn_watcher(tenant, settings.DEVICE_SUBJECT, DeviceHandler())
+                TenancyHandler._spawn_watcher(tenant, settings.DATA_SUBJECT, DataHandler(tenant))
+        else:
+            LOGGER.error('Failed to retrieve list of existing tenants (%d)', response.status_code)
+            exit(1)
+
+        return None
+
+    def handle_event(self, message):
+        """
+            Given a tenancy livecyle event, spawn the needed consumers to hanle its data
+            :param message Tenant lifecyle message as produced by Auth
+        """
+        data = json.loads(message)
+        LOGGER.debug('got tenancy event for tenant: %s', data['tenant'])
+
+        tenant = data['tenant']
+        TenancyHandler._spawn_watcher(tenant, settings.DEVICE_SUBJECT, DeviceHandler())
+        TenancyHandler._spawn_watcher(tenant, settings.DATA_SUBJECT, DataHandler(tenant))
+
+
+class DeviceHandler(KafkaEventHandler):
+    def __init__(self):
+        self.db = None
+
+    def handle_event(self, message):
+        """
+            Given a device management event, create (if not alredy existent) proper indexes
+            to suppor the new device
+            :param message Device lifecyle message, as produced by device manager
+        """
+        data = json.loads(message)
+        LOGGER.debug('got device event %s', message)
+
+        if self.db is None:
+            self.db = pymongo.MongoClient(settings.MONGO_DB, replicaSet=settings.replica_set)
+            self.db = self.db['device_history']
+
+        collection_name = "{}_{}".format(data['meta']['service'], data['data']['id'])
+        self.db[collection_name].create_index([('ts', pymongo.DESCENDING),
+                                               ('attr', pymongo.DESCENDING)],
+                                              unique=True)
+        self.db[collection_name].create_index('ts', expireAfterSeconds=settings.DATA_EXP)
```

Comment:
ts or @TS ?

Reply:
`ts` - the idea is to expire data based on their most accurate timestamp.

This is also a good opportunity to clarify something. This PR mainly handles the backport (for the new context broker, that is); while also backporting the feature needed by agrotracking.

The full-fledged agrotracking adaptation will be shortly published on its own fork (agrotracking/history-ws).